src/solvers/thermal_solver.hpp

```diff
@@ -110,43 +113,43 @@ class DynamicConductionOperator : public mfem::TimeDependentOperator
 protected:
 
   /// Finite Element space
-  mfem::ParFiniteElementSpace *m_fespace;
+  std::shared_ptr < mfem::ParFiniteElementSpace > m_fespace;
 
   /// Grid function for boundary condition projection
   mfem::ParGridFunction *m_state_gf;
 
   /// Solver for the mass matrix
-  mfem::CGSolver *m_M_solver;
+  std::shared_ptr < mfem::CGSolver > m_M_solver;
 
   /// Solver for the T matrix
-  mfem::CGSolver *m_T_solver;
+  std::shared_ptr < mfem::CGSolver > m_T_solver;
 
   /// Preconditioner for the M matrix
-  mfem::HypreSmoother *m_M_prec;
+  std::shared_ptr < mfem::HypreSmoother > m_M_prec;
 
   /// Preconditioner for the T matrix
-  mfem::HypreSmoother *m_T_prec;
+  std::shared_ptr < mfem::HypreSmoother > m_T_prec;
 
   /// Pointer to the assembled M matrix
-  mfem::HypreParMatrix *m_M_mat;
+  std::shared_ptr < mfem::HypreParMatrix > m_M_mat;
 
   /// Pointer to the eliminated M matrix
-  mfem::HypreParMatrix *m_M_e_mat;
+  std::shared_ptr < mfem::HypreParMatrix > m_M_e_mat;
 
   /// Pointer to the assembled K matrix
-  mfem::HypreParMatrix *m_K_mat;
+  std::shared_ptr < mfem::HypreParMatrix > m_K_mat;
 
   /// Pointer to the eliminated K matrix
-  mfem::HypreParMatrix *m_K_e_mat;
+  std::shared_ptr < mfem::HypreParMatrix > m_K_e_mat;
 
   /// Pointer to the assembled T ( = M + dt K) matrix
-  mfem::HypreParMatrix *m_T_mat;
+  std::shared_ptr < mfem::HypreParMatrix > m_T_mat;
 
   /// Pointer to the eliminated T matrix
-  mfem::HypreParMatrix *m_T_e_mat;
+  std::shared_ptr < mfem::HypreParMatrix > m_T_e_mat;
 
   /// Assembled RHS vector
-  mfem::Vector *m_rhs;
+  std::shared_ptr < mfem::Vector > m_rhs;
 
   /// RHS vector including essential boundary elimination
   mfem::Vector *m_bc_rhs;
```

Comment:
I assume the raw pointer is needed here due to interactions with MFEM?

Reply:
Nah, I was just on the fence about reworking the B.C. implementation. In hindsight though, it makes more sense to break those up into separate changes (e.g. move to smart pointers, then consider reorganizing boundary condition stuff). Hopefully I can get things building locally and finish inserting smart pointers into the rest of the code!